src/game/underwater_effects.c

```diff
@@ -0,0 +1,263 @@
+    transform->unk0 = 0;
+    transform->width = 0;
+    transform->height = 0;
+    transform->x = 0;
+    transform->y = 0;
+
+    return t;
+}
+
+// Called when air bubbles spawn underwater
+struct Task *SpawnAirBubbles(s32 p0, s32 p1, s32 p2, s32 p3)
+{
+    if ((s8)gSmallAirBubbleCount > 11) {
+        return NULL;
+    } else {
+        struct Task *t;
+        TaskStrc_801F15C *ts;
+        Sprite *s;
+        SpriteTransform *transform;
+
+        gSmallAirBubbleCount++;
+
+        t = sub_801F15C(0, 0, 0, 0, Task_SpawnAirBubbles,
+                        TaskDestructor_SpawnAirBubbles);
+
+        ts = TaskGetStructPtr(t);
+        s = &ts->s;
+        transform = &ts->transform;
+
+        ts->x = p0;
+        ts->y = p1;
+        ts->unk8 = p2;
+        ts->unkA = -0x80;
+        ts->unk10 = 0;
+        ts->unk12 = 0x100;
+        ts->unk14 = 0;
+
+        if (p3 == 0) {
+            s->graphics.dest = (void *)(OBJ_VRAM0 + 0x2A40);
+            s->graphics.anim = SA2_ANIM_BUBBLES_SMALL;
+            s->variant = 0;
+        } else {
+            s->graphics.dest = (void *)(OBJ_VRAM0 + 0x2AC0);
+            s->graphics.anim = SA2_ANIM_BUBBLES_GROUP;
+            s->variant = 0;
+
+            ts->unk14 = (((u32)PseudoRandom32() & 0x30000) >> 16);
+        }
+
+        s->unk1A = 0x240;
+
+        transform->unk0 = 0;
+        transform->width = 0;
+        transform->height = 0;
+        transform->x = 0;
+        transform->y = 0;
+
+        return t;
+    }
+}
+
+bool32 RandomlySpawnAirBubbles(Player *p)
+{
+    u32 rand, randX, randY;
+
+    u32 result = FALSE;
+
+    u32 flags = gUnknown_03005590 & 0x7;
+    if (!flags) {
+        rand = ((u32)PseudoRandom32() & 0x300);
+        if (!rand) {
+            randX = (((u32)PseudoRandom32() & 0xF00) >> 8);
+            randY = (((u32)PseudoRandom32() & 0xF00) >> 8) - 8;
+
+            if (!(p->moveState & MOVESTATE_FACING_LEFT))
+                randX = -randX;
+
+            SpawnAirBubbles(p->x - randX, p->y - randY, p->speedAirX,
+                            ((u32)PseudoRandom32() & 0x100) >> 8);
+
+            result = TRUE;
+        }
+    }
+
+    return result;
+}
+
+static void Task_SpawnAirBubbles(void)
+{
+    TaskStrc_801F15C *ts = TaskGetStructPtr(gCurTask);
+    Sprite *s = &ts->s;
+    SpriteTransform *transform = &ts->transform;
+
+    s32 r1 = ts->x;
+    s32 r4 = ts->y;
+    s32 r2;
+    s32 unk10 = ts->unk10;
+
+    r1 += SIN((unk10 & 0xFF) * 4) >> 4;
+    {
+        struct Camera *cam = &gCamera;
+        transform->x = Q_24_8_TO_INT(r1) - cam->x;
+        transform->y = Q_24_8_TO_INT(r4) - cam->y;
+    }
+
+    r2 = ((unk10 + 1) << 4);
+    r2 = MIN(r2, 0x100);
+
+    transform->width = r2;
+    transform->height = r2;
+
+    if (ts->unk14 & 0x1)
+        transform->width = -r2;
+
+    if (ts->unk14 & 0x2)
+        transform->height = -transform->height;
+
+    if (((u16)(transform->x + 0x20) > 0x130) || ((u16)(transform->y + 0x20) > 0xE0)
```

Comment:
Can be converted to a `x > -0x20 && x < 0x110`? Or whatever, I can't remember the exact way to do it

Reply:
Oh right! But I don't know how to match that rn either... will leave it for now!